assets/src/components/editing/markdown_editor/MarkdownEditor.tsx

```diff
@@ -0,0 +1,115 @@
+import { debounce } from 'lodash';
+import { Descendant } from 'slate';
+import { Icon } from 'components/misc/Icon';
+import { NormalizerContext } from '../editor/normalizers/normalizer';
+import { CommandContext } from '../elements/commands/interfaces';
+import { contentMarkdownDeserializer } from './content_markdown_deserializer';
+import { serializeMarkdown } from './content_markdown_serializer';
+
+interface MarkdownEditorProps {
+  onEdit: (value: Descendant[], _editor: any, _operations: any[]) => void;
+  onSwitchModes: () => void;
+  // The content to display
+  value: Descendant[];
+  // Whether or not editing is allowed
+  editMode: boolean;
+  fixedToolbar?: boolean;
+  commandContext: CommandContext;
+  normalizerContext?: NormalizerContext;
+  className?: string;
+  style?: React.CSSProperties;
+  placeholder?: string;
+  initialHeight?: number;
+  onPaste?: React.ClipboardEventHandler<HTMLDivElement>;
+  children?: React.ReactNode;
+  onFocus?: FocusEventHandler | undefined;
+  onBlur?: FocusEventHandler | undefined;
+}
+
+export const MarkdownEditor: React.FC<MarkdownEditorProps> = (props) => {
+  const [value, setValue] = useState<string | undefined>(() =>
+    contentMarkdownDeserializer(props.value),
+  );
+  const [lastSavedValue, setLastSavedValue] = useState<string | undefined>();
+
+  const switchToSlateCommand: ICommand = {
+    name: 'Switch to Slate',
+    keyCommand: 'switch-to-slate',
+    buttonProps: { 'aria-label': 'Switch to Slate' },
+    icon: <Icon icon="newspaper" />,
+    execute: () => {
+      props.onSwitchModes();
+      const content = serializeMarkdown(value || '');
+      props.onEdit(content as Descendant[], null, []); // Trigger a save as well
+    },
+  };
+
+  const darkMode: boolean = useMemo(() => {
+    return document.documentElement.classList.contains('dark');
+  }, []);
+  const { onEdit, initialHeight } = props;
+  const modeClass = darkMode ? 'dark' : 'light';
+
+  const saveChanges = useCallback(
+    (newValue: string) => {
+      if (newValue === lastSavedValue) {
+        return;
+      }
+      setLastSavedValue(newValue);
+      const content = serializeMarkdown(newValue);
+      onEdit(content as Descendant[], null, []);
+    },
+    [lastSavedValue, onEdit],
+  );
+
+  const delayedSaveChanges = useMemo(
+    () => debounce(saveChanges, 800, { maxWait: 5000 }),
+    [saveChanges],
+  );
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const onChange = useCallback(
+    (newValue: string | undefined) => {
+      console.info('onchange');
+      setValue(newValue || '');
+      delayedSaveChanges(newValue || '');
+    },
+    [delayedSaveChanges],
+  );
+
+  const onBlur = useCallback(() => {
+    console.info('onblur');
+    saveChanges(value || '');
+  }, [value, saveChanges]);
+
+  return (
+    <MDEditor
+      value={value}
+      className={props.className}
+      onChange={onChange}
+      height={initialHeight || 600}
```

Comment:
Is there any way to allow to have the height of the `MDEditor` component be driven from the content that is actually present?  It seems like each MDEditor is fixed to a certain height, and then internally scrolls the content through that fixed viewport.  It would be strongly preferred if the height of this thing could React to the content that is present, so that the author never has to scroll and never is wasting UI real estate (when they have only a few lines of content).

Reply:
I just made this change, it is a big improvement. Also got rid of the trailing \n\n in each editor's content when it first starts, makes that editing slightly easier and worked well with the resize change.

![image](https://github.com/Simon-Initiative/oli-torus/assets/333265/77d333f3-f37b-48e9-8c50-621d1894aa39)

![image](https://github.com/Simon-Initiative/oli-torus/assets/333265/080a70be-3e7e-4116-ab9f-d6b0ae109985)